### DISCUSSION TITLE
How to add item to ItemsPanelTemplate using C# Markup?

### DISCUSSION THREAD

**Discussion by kucint** [unoplatform/uno#14947] [Q&A] [answered]
I intend to create a user control that presents a collection of some elements on the canvas.
In XAML this would be realized by following (pseudo) code:

```
<UserControl>
  <ScrollViewer>
    <ItemsControl>
      <ItemsControl.ItemsPanel>
        <ItemsPanelTemplate>
          <Canvas/>
        </ItemsPanelTemplate>
      </ItemsControl.ItemsPanel>
      <ItemsControl.ItemContainerStyle>
        <Style TargetType="ContentPresenter">
          <Setter Property="internal:CanvasBindingHelper.CanvasLeftBindingPath" Value="Pos.X"/>
          <Setter Property="internal:CanvasBindingHelper.CanvasTopBindingPath" Value="Pos.Y"/>
        </Style>
      </ItemsControl.ItemContainerStyle>
    </ItemsControl>
  </ScrollViewer>
</UserControl>
```

I try to do the same but this time using C# markup:

```
public sealed partial class Pipeline : UserControl
{
    public Pipeline()
    {
        this.Content(
                new ScrollViewer()
                .Content(
                    new ItemsControl()
                    .ItemsPanel(
                        new ItemsPanelTemplate()
                        {
                            // I am stuck here: how to create Canvas as a child of ItemsPanelTemplate?
                        }
                    )
                )
            );
    }
}
```

How should I create Canvas as a child of ItemsPanelTemplate?
thanks for help!
tomasz

**Comment by dansiegel** [comment, marked as answer]
The ItemsPanelTemplate is just a type of FrameworkTemplate. Like all FrameworkTemplates extensions are created in C# Markup to provide a delegate to create the UI.

```cs
public sealed partial class Pipeline : UserControl
{
    public Pipeline()
    {
        this.Content(
            new ScrollViewer()
            .Content(
                new ItemsControl()
                .ItemsPanel(() => new Canvas())
            )
        );
    }
}
```

**Reply by kucint** [reply]
Hello @dansiegel 

Unfortunately, something is wrong here.
`new ItemsControl().ItemsPanel(() => new Canvas())` seems to not work as expected; the container is still a vertical `StackLayout` (which is default container) rather than `Canvas` that has not been created.

**Question 1**: How correctly create Canvas as Container of `ItemsPanel`?

You can see on screenshot below that `StackLayout` is still in action:

![image](https://github.com/unoplatform/uno/assets/54866088/bad925b3-4014-4565-aaff-10b1fcd1aae5)

Have a look on a code snippet:

```
public sealed partial class PipelineEditor : UserControl
{
    public PipelineEditor(PipelineModel model) => this
        .DataContext(model, (ctrl, vm) => ctrl
        .Background(ThemeResource.Get<Brush>("ApplicationPageBackgroundThemeBrush"))
        .Content(ContentB())
        );

    //  this works fine, it creates Canvas directly and all elements are correctly positioned
    private static UIElement ContentA()
        => new Canvas()
        .Width(800)
        .Height(600)
        .Background(Colors.LightGray)
        .Children(
                       new Rectangle()
                       .Fill(Colors.White)
                       .Stroke(Colors.Black)
                       .StrokeThickness(1)
                       .Width(50)
                       .Height(50)
                       .Canvas(left: 100, top: 100),

                       new Rectangle()
                       .Fill(Colors.DarkKhaki)
                       .Stroke(Colors.Black)
                       .StrokeThickness(1)
                       .Width(50)
                       .Height(50)
                       .Canvas(left: 200, top: 100)
                       );


    //  this doesn't work as expected, Canvas is not created, StackPanel is still in action
    private static UIElement ContentB()
        => new ItemsControl()
        .ItemsPanel(() => new Canvas())
        .ItemsSource(new Node[]
        {
            new ("A", new SolidColorBrush(Colors.DarkMagenta)),
            new ("B", new SolidColorBrush(Colors.BlanchedAlmond)),
            new ("C", new SolidColorBrush(Colors.HotPink))
        })
        .ItemTemplate<Node>(
            item => new Rectangle()
                .Width(100).Height(50)
                .Stroke(Colors.LightGray)
                .Fill(() => item.Brush)               // using binding rather than direct value which is always null
                .Canvas(left: 100, top: 100)      // positioning not working, see **Question 3**
            );
}

public sealed record Node(string Label, Brush Brush);
```

**Question 2:** How do I set a background color of a `Canvas`?

if I would like to modify the container, I expect to do the following:
```
    private static UIElement ContentB()
        => new ItemsControl().ItemsPanel(() => new Canvas().Background(Colors.LightGray));
```
Unfortunately, this doesn't work, it causes app crash:
_Exception thrown: 'System.ArgumentException' in WinRT.Runtime.dll
An exception of type 'System.ArgumentException' occurred in WinRT.Runtime.dll but was not handled in user code_

The equivalent Code in XAML would be:
```
      <ItemsPanelTemplate>
        <Canvas Background="Transparent"/>
      </ItemsPanelTemplate>
    </ItemsControl.ItemsPanel>
```
How do I realize it in C# Markup?

**Question 3:** How do I a deal in C# Markup with known issue that elements on `Canvas` does not reflect position correctly until a custom canvas binding is provided?
In XAML, one must provide following code:
```
      <ItemsControl.ItemContainerStyle>
        <Style TargetType="ContentPresenter">
          <Setter Property="internal:CanvasBindingHelper.CanvasLeftBindingPath" Value="Pos.X"/>
          <Setter Property="internal:CanvasBindingHelper.CanvasTopBindingPath" Value="Pos.Y"/>
        </Style>
      </ItemsControl.ItemContainerStyle>
```
what is equivalent in C# Markup?

thank you for your help!

**Reply by dansiegel** [reply]
@kucint sorry I had missed this... 

1) If it's not behaving the way I described in my answer above I would encourage you to create a reproduction and open an issue on it.
2) What you are showing is exactly correct and you should not be getting an exception. If you could please provide a reproduction and open an issue we would be happy to look into it further.
3) Please see: unoplatform/uno.extensions#2218 & the [Styles docs](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Markup/Styles.html)

**Reply by kazo0** [reply]
@kucint I can confirm for **Question 1** that the `ItemsPanel` doesn't seem to be changing and I have created a repro app and issue to track this here: https://github.com/unoplatform/uno/issues/16046

**Comment by kucint** [comment]
thanks @dansiegel .
I missed it somehow.

**Reply by sasakrsmanovic** [reply]
can this discussion be closed and marked as 'answered'? @kucint

**Reply by kucint** [reply]
Hi @sasakrsmanovic 
not really, I posted recently 3 questions, that still need to be answered: https://github.com/unoplatform/uno/discussions/14947#discussioncomment-8852296

**Comment by francoistanguay** [comment]
@kucint I think it would be better if this issues with 3 questions was split into 3 discussions for easier tracking

**Reply by kucint** [reply]
I created 3 independent discussions here:

- https://github.com/unoplatform/uno/issues/16090
- https://github.com/unoplatform/uno/discussions/16091
- https://github.com/unoplatform/uno/discussions/16092
